src/contracts/StakingChef.sol

```diff
@@ -95,73 +77,72 @@ contract StakingChef is Chef {
         setLastRewardBlock(block.number);
     }
 
-// --------------------------------------------------------------------- Manage rewards and Principal
+    // --------------------------------------------------------------------- Manage rewards and Principal
 
-    function claim(uint256 stakeIndex) public {
-        Staker storage stake = _stakes[stakeIndex];
-        harvest(stakeIndex);
+    function claim() public {
+        Staker storage stake = staker[msg.sender];
+        harvest();
 
         require(stake.claimable != 0, "No claimable rewards to withdraw");
-        require(stake.stakeAddress == _msgSender(), "Sender can not claim");
 
         SafeERC20.safeTransfer(IERC20(getMetricToken()), msg.sender, stake.claimable);
         stake.claimable = 0;
 
-        emit Withdraw(msg.sender, stakeIndex, stake.claimable);
+        emit TransferPrincipal(msg.sender, stake, stake.claimable);
     }
 
-    function withdrawPrincipal(uint256 stakeIndex) public {
-        Staker storage stake = _stakes[stakeIndex];
-
+    function withdrawPrincipal() public {
+        Staker storage stake = staker[msg.sender];
         require(stake.metricAmount != 0, "No Metric to withdraw");
-        require(stake.stakeAddress == _msgSender(), "Sender can not withdraw");
 
         SafeERC20.safeTransfer(IERC20(getMetricToken()), msg.sender, stake.metricAmount);
         stake.metricAmount = 0;
 
-        emit Withdraw(msg.sender, stakeIndex, stake.metricAmount);
+        emit TransferPrincipal(msg.sender, stake, stake.metricAmount);
     }
 
-    function harvest(uint256 stakeIndex) internal {
-        Staker storage stake = _stakes[stakeIndex];
-
+    function harvest() internal {
+        Staker storage stake = staker[msg.sender];
```

Comment:
if this is called internally I don't believe it has access to msg.sender - I think you would need to pass it in from whatever is invoking it

Reply:
I think I maybe be wrong about this  - in sushiswap an internal function has access to msg.sender

Reply:
it will work -  the other option is `tx.origin` which we do not want to use - usually ever.   It'll become more and more common for people to write their contracts to call into ours, and we should work as expected when that happens.